src/main/scala/org/neo4j/spark/writer/Neo4jDataWriter.scala

```diff
@@ -0,0 +1,115 @@
+package org.neo4j.spark.writer
+
+import java.util
+import java.util.Collections
+import java.util.concurrent.CountDownLatch
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SaveMode
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.sources.v2.writer.{DataWriter, WriterCommitMessage}
+import org.apache.spark.sql.types.StructType
+import org.neo4j.driver.exceptions.{Neo4jException, ServiceUnavailableException, SessionExpiredException}
+import org.neo4j.driver.{Session, Transaction, Values}
+import org.neo4j.spark.service.{MappingService, Neo4jQueryService, Neo4jQueryWriteStrategy}
+import org.neo4j.spark.util.Neo4jUtil._
+import org.neo4j.spark.util.Neo4jImplicits._
+import org.neo4j.spark.util.Neo4jUtil
+import org.neo4j.spark.{DriverCache, Neo4jOptions}
+
+class Neo4jDataWriter(jobId: String,
+                      partitionId: Int,
+                      structType: StructType,
+                      saveMode: SaveMode,
+                      options: Neo4jOptions) extends DataWriter[InternalRow] with Logging {
+
+  private val driverCache: DriverCache = new DriverCache(options.connection, jobId)
+
+  private var transaction: Transaction = _
+  private var session: Session = _
+
+  private val mappingService = new MappingService(options)
+
+  private val batch: util.List[java.util.Map[String, Object]] = new util.ArrayList[util.Map[String, Object]]()
+
+  private val retries = new CountDownLatch(options.transactionMetadata.retries)
+
+  val query: String = new Neo4jQueryService(options, new Neo4jQueryWriteStrategy(saveMode)).createQuery()
+
+  override def write(record: InternalRow): Unit = {
+    batch.add(mappingService.toParameter(record, structType))
+    if (batch.size() == options.nodeMetadata.batchSize) {
+      writeBatch
```

Comment:
docs need reference: since fundamentally we're doing batched writes, if a late batch fails, this isn't transactional and all the previous batches still stay

Reply:
we can improve this in the future by adding a property like `batch.commit=3` that says commit every three batches (= 5k*3=15k rows in the default case). WDYT?

Reply:
and we can also provide an `ingestion.mode=BATCH/TRANSACTIONAL` that can allow the use to choose the ingestion mode. WDYT?

Reply:
Fine point, but if you put 3 batches in a transaction, I think it still requires the memory of 3 batches.  So I don't see the point in putting 3 batches in 1 TX, why not just make 1 bigger batch.  The limiting factor is memory and the potential for the server to OOM.

As for transactional, it seems to me you don't need this mode, because it's the same as making the batch size more than the total number of records in the write.   I.e. make the entire write one batch.

I think we need to document how it works, but basic batching with an adjustable size is the right way to go.  It's simple, easy to implement, and also easy to understand.  If users don't like batching and want it to be transactional, they can just make the batch number really really large assuming they have the memory.